### PR TITLE
JDK-8300240: Replace NULL with nullptr in share/ci/

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory share/ci/. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300240](https://bugs.openjdk.org/browse/JDK-8300240): Replace NULL with nullptr in share/ci/

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12029/head:pull/12029` \
`$ git checkout pull/12029`

Update a local copy of the PR: \
`$ git checkout pull/12029` \
`$ git pull https://git.openjdk.org/jdk pull/12029/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12029`

View PR using the GUI difftool: \
`$ git pr show -t 12029`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12029.diff">https://git.openjdk.org/jdk/pull/12029.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/ci/ciEnv.cpp [hunk 1]
```diff
@@ -1354,7 +1354,7 @@ bool ciEnv::print_dyno_loc(outputStream* out, const InstanceKlass* ik) const {
 }
 
 // Look up the location descriptor for the given class and return it as a string.
-// Returns NULL if no location is found.
+// Returns nullptr if no location is found.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciEnv.cpp:1357]: fix

File: src/hotspot/share/ci/ciEnv.hpp [hunk 2]
```diff
@@ -107,7 +107,7 @@ class ciEnv : StackObj {
   ciInstance* _the_min_jint_string; // The Java string "-2147483648"
 
   // Look up a klass by name from a particular class loader (the accessor's).
-  // If require_local, result must be defined in that class loader, or NULL.
+  // If require_local, result must be defined in that class loader, or nullptr.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciEnv.hpp:110]: fix

File: src/hotspot/share/ci/ciEnv.hpp [hunk 3]
```diff
@@ -269,7 +269,7 @@ class ciEnv : StackObj {
   }
 
   // See if we already have an unloaded klass for the given name
-  // or return NULL if not.
+  // or return nullptr if not.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciEnv.hpp:272]: fix

File: src/hotspot/share/ci/ciField.cpp [hunk 4]
```diff
@@ -44,7 +44,7 @@
 // the ciField will be incomplete.
 
 // The ciObjectFactory cannot create circular data structures in one query.
-// To avoid vicious circularities, we initialize ciField::_type to NULL
+// To avoid vicious circularities, we initialize ciField::_type to nullptr
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciField.cpp:47]: fix

File: src/hotspot/share/ci/ciInstance.cpp [hunk 5]
```diff
@@ -42,16 +42,16 @@
 ciType* ciInstance::java_mirror_type() {
   VM_ENTRY_MARK;
   oop m = get_oop();
-  // Return NULL if it is not java.lang.Class.
-  if (m == NULL || m->klass() != vmClasses::Class_klass()) {
-    return NULL;
+  // Return nullptr if it is not java.lang.Class.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciInstance.cpp:45]: fix

File: src/hotspot/share/ci/ciInstanceKlass.hpp [hunk 6]
```diff
@@ -72,7 +72,7 @@ class ciInstanceKlass : public ciKlass {
   int                    _has_injected_fields; // any non static injected fields? lazily initialized.
 
   // The possible values of the _implementor fall into following three cases:
-  //   NULL: no implementor.
+  //   nullptr: no implementor.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciInstanceKlass.hpp:75]: fix

File: src/hotspot/share/ci/ciMethod.cpp [hunk 7]
```diff
@@ -659,7 +659,7 @@ bool ciMethod::parameter_profiled_type(int i, ciKlass*& type, ProfilePtrKind& pt
 // ciMethod::find_monomorphic_target
 //
 // Given a certain calling environment, find the monomorphic target
-// for the call.  Return NULL if the call is not monomorphic in
+// for the call.  Return nullptr if the call is not monomorphic in
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciMethod.cpp:662]: fix

File: src/hotspot/share/ci/ciMethod.cpp [hunk 8]
```diff
@@ -748,11 +748,11 @@ ciMethod* ciMethod::find_monomorphic_target(ciInstanceKlass* caller,
     // If we are going to reason about inheritance, it's easiest
     // if the method in question is public, protected, or private.
     // If the answer is not root_m, it is conservatively correct
-    // to return NULL, even if the CHA encountered irrelevant
+    // to return nullptr, even if the CHA encountered irrelevant
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciMethod.cpp:751]: fix

File: src/hotspot/share/ci/ciMethod.cpp [hunk 9]
```diff
@@ -810,9 +810,9 @@ ciMethod* ciMethod::resolve_invoke(ciKlass* caller, ciKlass* exact_receiver, boo
     }
   }
 
-  if (m == NULL) {
-    // Return NULL only if there was a problem with lookup (uninitialized class, etc.)
-    return NULL;
+  if (m == nullptr) {
+    // Return nullptr only if there was a problem with lookup (uninitialized class, etc.)
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciMethod.cpp:814]: fix

File: src/hotspot/share/ci/ciMethod.hpp [hunk 10]
```diff
@@ -283,20 +283,20 @@ class ciMethod : public ciMetadata {
...
     ciSignature* declared_signature;
     get_method_at_bci(bci, ignored_will_link, &declared_signature);
-    assert(declared_signature != NULL, "cannot be null");
+    assert(declared_signature != nullptr, "cannot be null");
     return declared_signature;
   }
 
   // Given a certain calling environment, find the monomorphic target
-  // for the call.  Return NULL if the call is not monomorphic in
+  // for the call.  Return nullptr if the call is not monomorphic in
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciMethod.hpp:291]: fix

File: src/hotspot/share/ci/ciMethod.hpp [hunk 11]
```diff
@@ -283,20 +283,20 @@ class ciMethod : public ciMetadata {
...
   // its calling environment.
   ciMethod* find_monomorphic_target(ciInstanceKlass* caller,
                                     ciInstanceKlass* callee_holder,
                                     ciInstanceKlass* actual_receiver,
                                     bool check_access = true);
 
   // Given a known receiver klass, find the target for the call.
-  // Return NULL if the call has no target or is abstract.
+  // Return nullptr if the call has no target or is abstract.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciMethod.hpp:299]: fix

File: src/hotspot/share/ci/ciMethodData.hpp [hunk 12]
```diff
@@ -403,13 +403,13 @@ class ciMethodData : public ciMetadata {
   // Coherent snapshot of original header.
   MethodData::CompilerCounters _orig;
 
-  // Area dedicated to parameters. NULL if no parameter profiling for this method.
+  // Area dedicated to parameters. nullptr if no parameter profiling for this method.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciMethodData.hpp:406]: fix

File: src/hotspot/share/ci/ciMethodData.hpp [hunk 13]
```diff
@@ -511,16 +511,16 @@ class ciMethodData : public ciMetadata {
...
+  bool is_valid(DataLayout* current)    { return current != nullptr; }
 
   DataLayout* extra_data_base() const  { return data_layout_at(data_size()); }
   DataLayout* args_data_limit() const  { return data_layout_at(data_size() + extra_data_size() -
                                                                parameters_size()); }
 
-  // Get the data at an arbitrary bci, or NULL if there is none. If m
-  // is not NULL look for a SpeculativeTrapData if any first.
-  ciProfileData* bci_to_data(int bci, ciMethod* m = NULL);
+  // Get the data at an arbitrary bci, or nullptr if there is none. If m
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciMethodData.hpp:521]: fix

File: src/hotspot/share/ci/ciNullObject.hpp [hunk 14]
```diff
@@ -22,8 +22,8 @@
  *
  */
 
-#ifndef SHARE_CI_CINULLOBJECT_HPP
-#define SHARE_CI_CINULLOBJECT_HPP
+#ifndef SHARE_CI_CInullptrOBJECT_HPP
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciNullObject.hpp:25]: fix

File: src/hotspot/share/ci/ciObjArrayKlass.cpp [hunk 15]
```diff
@@ -79,8 +79,8 @@ ciObjArrayKlass::ciObjArrayKlass(ciSymbol* array_name,
 //
 // What is the one-level element type of this array?
 ciKlass* ciObjArrayKlass::element_klass() {
-  if (_element_klass == NULL) {
-    assert(dimension() > 1, "_element_klass should not be NULL");
+  if (_element_klass == nullptr) {
+    assert(dimension() > 1, "_element_klass should not be nullptr");
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciObjArrayKlass.cpp:83]: fix

File: src/hotspot/share/ci/ciObject.cpp [hunk 16]
```diff
@@ -83,19 +83,19 @@ ciObject::ciObject(Handle h) {
...
+  assert(klass != nullptr, "must supply klass");
+  _handle = nullptr;
   _klass = klass;
 }
 
 // ------------------------------------------------------------------
 // ciObject::ciObject
 //
-// NULL variant.  Used only by ciNullObject.
+// nullptr variant.  Used only by ciNullObject.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciObject.cpp:94]: fix

File: src/hotspot/share/ci/ciObject.cpp [hunk 17]
```diff
@@ -111,14 +111,14 @@ oop ciObject::get_oop() const {
 //
 // Get the ciKlass of this ciObject.
 ciKlass* ciObject::klass() {
-  if (_klass == NULL) {
-    if (_handle == NULL) {
-      // When both _klass and _handle are NULL, we are dealing
+  if (_klass == nullptr) {
+    if (_handle == nullptr) {
+      // When both _klass and _handle are nullptr, we are dealing
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciObject.cpp:116]: fix

File: src/hotspot/share/ci/ciObject.hpp [hunk 18]
```diff
@@ -53,7 +53,7 @@ class ciObject : public ciBaseObject {
 
 private:
   // A JNI handle referring to an oop in the VM.  This
-  // handle may, in a small set of cases, correctly be NULL.
+  // handle may, in a small set of cases, correctly be nullptr.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciObject.hpp:56]: fix

File: src/hotspot/share/ci/ciTypeFlow.cpp [hunk 19]
```diff
@@ -588,7 +588,7 @@ void ciTypeFlow::StateVector::do_checkcast(ciBytecodeStream* str) {
   bool will_link;
   ciKlass* klass = str->get_klass(will_link);
   if (!will_link) {
-    // VM's interpreter will not load 'klass' if object is NULL.
+    // VM's interpreter will not load 'klass' if object is nullptr.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciTypeFlow.cpp:591]: fix

File: src/hotspot/share/ci/ciTypeFlow.hpp [hunk 20]
```diff
@@ -174,19 +174,19 @@ class ciTypeFlow : public ArenaObj {
       T_BOTTOM  = T_CONFLICT,
       T_LONG2   = T_SHORT,     // 2nd word of T_LONG
       T_DOUBLE2 = T_CHAR,      // 2nd word of T_DOUBLE
-      T_NULL    = T_BYTE       // for now.
+      T_nullptr    = T_BYTE       // for now.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciTypeFlow.hpp:177]: fix

File: src/hotspot/share/ci/ciTypeFlow.hpp [hunk 21]
```diff
@@ -174,19 +174,19 @@ class ciTypeFlow : public ArenaObj {
...
     };
     static ciType* top_type()    { return ciType::make((BasicType)T_TOP); }
     static ciType* bottom_type() { return ciType::make((BasicType)T_BOTTOM); }
     static ciType* long2_type()  { return ciType::make((BasicType)T_LONG2); }
     static ciType* double2_type(){ return ciType::make((BasicType)T_DOUBLE2); }
-    static ciType* null_type()   { return ciType::make((BasicType)T_NULL); }
+    static ciType* null_type()   { return ciType::make((BasicType)T_nullptr); }
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciTypeFlow.hpp:183]: fix

File: src/hotspot/share/ci/ciTypeFlow.hpp [hunk 22]
```diff
@@ -333,16 +333,16 @@ class ciTypeFlow : public ArenaObj {
     }
     // pop_objArray and pop_typeArray narrow the tos to ciObjArrayKlass
     // or ciTypeArrayKlass (resp.).  In the rare case that an explicit
-    // null is popped from the stack, we return NULL.  Caller beware.
+    // null is popped from the stack, we return nullptr.  Caller beware.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciTypeFlow.hpp:336]: fix

File: src/hotspot/share/ci/ciTypeFlow.hpp [hunk 23]
```diff
@@ -818,11 +818,11 @@ class ciTypeFlow : public ArenaObj {
   int backedge_copy_count(int ciBlockIndex, JsrSet* jsrs) const;
 
   // Return an existing block containing bci which has a JsrSet compatible
-  // with jsrs, or NULL if there is none.
+  // with jsrs, or nullptr if there is none.
```
Comment (jdksjolen) [inline src/hotspot/share/ci/ciTypeFlow.hpp:821]: fix

**Review by jdksjolen — COMMENTED** [review]
Manual fixes to be done.

**Comment by jdksjolen** [comment]
All of the issues I found are now fixed and this is open for review.

File: src/hotspot/share/ci/ciInstance.cpp [hunk 24]
```diff
@@ -42,16 +42,16 @@
 ciType* ciInstance::java_mirror_type() {
   VM_ENTRY_MARK;
   oop m = get_oop();
-  // Return NULL if it is not java.lang.Class.
-  if (m == NULL || m->klass() != vmClasses::Class_klass()) {
-    return NULL;
+  // Return null if it is not java.lang.Class.
+  if (m == nullptr || m->klass() != vmClasses::Class_klass()) {
+    return nullptr;
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciInstance.cpp:47]: I am confused. Did I miss some previous reviews which asked to keep `null` in comments?
These lines are confusing: comment says `null` but method returns `nullptr`. Why not use `nullptr` in comment?

File: src/hotspot/share/ci/ciInstance.hpp [hunk 25]
```diff
@@ -53,7 +53,7 @@ class ciInstance : public ciObject {
 
 public:
   // If this object is a java mirror, return the corresponding type.
-  // Otherwise, return NULL.
+  // Otherwise, return nullptr.
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciInstance.hpp:56]: And here we have `nullptr` in comment.

File: src/hotspot/share/ci/ciMethodData.hpp [hunk 26]
```diff
@@ -511,16 +511,16 @@ class ciMethodData : public ciMetadata {
...
   DataLayout* extra_data_base() const  { return data_layout_at(data_size()); }
   DataLayout* args_data_limit() const  { return data_layout_at(data_size() + extra_data_size() -
                                                                parameters_size()); }
 
-  // Get the data at an arbitrary bci, or NULL if there is none. If m
-  // is not NULL look for a SpeculativeTrapData if any first.
-  ciProfileData* bci_to_data(int bci, ciMethod* m = NULL);
+  // Get the data at an arbitrary bci, or null if there is none. If m
+  // is not nullptr look for a SpeculativeTrapData if any first.
+  ciProfileData* bci_to_data(int bci, ciMethod* m = nullptr);
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciMethodData.hpp:523]: Again mixture of `null` and `nullptr` in comment. Even so the comment for `ciMethodData::bci_to_data()` in .cpp file says `nullptr`

File: src/hotspot/share/ci/ciNullObject.hpp [hunk 27]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1999, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1999, 2023, Oracle and/or its affiliates. All rights reserved.
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciNullObject.hpp:2]: No change in file

**Comment by jdksjolen** [comment]
Hi @vnkozlov.

Regarding the confusion regarding `null` and `nullptr` in comments :). Yes, the idea is that we use `null` in comments as it reads better. In *some* situations I've chosen to keep `nullptr` in comments, this is specifically when there are code expressions (such as `nullptr == x`) in the comments. Sometimes I simply miss fixing them, because my script cannot deal with comments correctly and I need to alter them manually. This fact makes it a bit confusing to review, I agree.

This was discussed in a previous review, as you suspected.

Thank you for your comments!

**Comment by vnkozlov** [comment]
> Regarding the confusion regarding `null` and `nullptr` in comments :). Yes, the idea is that we use `null` in comments as it reads better. In _some_ situations I've chosen to keep `nullptr` in comments, this is specifically when there are code

As owner of the code I prefer to match comments and code.

> This was discussed in a previous review, as you suspected.

Please, point me to that discussion.

File: src/hotspot/share/ci/ciInstance.hpp (hunk 25 above)
Comment (vnkozlov) [reply src/hotspot/share/ci/ciInstance.hpp:56]: Use `null` here.

File: src/hotspot/share/ci/ciMethod.cpp [hunk 28]
```diff
@@ -783,7 +783,7 @@ bool ciMethod::can_omit_stack_trace() const {
 // ciMethod::resolve_invoke
 //
 // Given a known receiver klass, find the target for the call.
-// Return NULL if the call has no target or the target is abstract.
+// Return nullptr if the call has no target or the target is abstract.
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciMethod.cpp:786]: Use `null`

File: src/hotspot/share/ci/ciMethod.cpp [hunk 29]
```diff
@@ -1032,11 +1032,11 @@ ciMethodData* ciMethod::method_data() {
 // ------------------------------------------------------------------
 // ciMethod::method_data_or_null
 // Returns a pointer to ciMethodData if MDO exists on the VM side,
-// NULL otherwise.
+// nullptr otherwise.
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciMethod.cpp:1035]: Use `null`

File: src/hotspot/share/ci/ciMethodData.hpp [hunk 30]
```diff
@@ -511,16 +511,16 @@ class ciMethodData : public ciMetadata {
...
 
   DataLayout* extra_data_base() const  { return data_layout_at(data_size()); }
   DataLayout* args_data_limit() const  { return data_layout_at(data_size() + extra_data_size() -
                                                                parameters_size()); }
 
-  // Get the data at an arbitrary bci, or NULL if there is none. If m
-  // is not NULL look for a SpeculativeTrapData if any first.
-  ciProfileData* bci_to_data(int bci, ciMethod* m = NULL);
+  // Get the data at an arbitrary bci, or null if there is none. If m
+  // is not nullptr look for a SpeculativeTrapData if any first.
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciMethodData.hpp:522]: Use `null`

File: src/hotspot/share/ci/ciNullObject.hpp (hunk 27 above)
Comment (vnkozlov) [reply src/hotspot/share/ci/ciNullObject.hpp:2]: Undo this. No changes in file.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Okay. After second look through I agree to use `null` in comments as it indeed sounds better.
I found several cases with `nullptr` in comments you missed. We should be consistent.
And one file which has copyright year update but no changes.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by jdksjolen** [comment]
Passes tier1.

**Review by coleenp — APPROVED** [review]

**Comment by jdksjolen** [comment]
Still passes tier1!

/integrate

**Pull request closed without merging**